### ISSUE TITLE
torch.compile doesn't work with custom __getitem__ in model

### ISSUE THREAD

**Issue by nisheethlahoti** [pytorch/pytorch#91432]
### 🐛 Describe the bug

The following file, on being run as main (with nightly pytorch 2.0dev), throws an AssertionError:
```python
import torch
from torch import Tensor, nn


class BufferList(nn.Module):
    """Like torch.nn.ParameterList, but for buffers instead of parameters"""

    def __init__(self, *buffers: Tensor | None) -> None:
        super().__init__()
        for x in buffers:
            self.register_buffer(str(len(self._buffers)), x)

    def __getitem__(self, idx: int) -> Tensor | None:
        return self._buffers[str(idx + len(self._buffers) if idx < 0 else idx)]

    def __len__(self) -> int:
        return len(self._buffers)


class Dummy(nn.Module):
    def __init__(self) -> None:
        super().__init__()
        self.buflist = BufferList(torch.arange(5.0))

    def forward(self, x: Tensor) -> Tensor:
        return x + self.buflist[0]

if __name__ == "__main__":
    x = torch.ones(5)
    model = torch.compile(Dummy())
    print(model(x))
```

Full error:
```
Traceback (most recent call last):
  File "/tmp/model.py", line 31, in <module>
    print(model(x))
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1482, in _call_impl
    return forward_call(*args, **kwargs)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 83, in forward
    return self.dynamo_ctx(self._orig_mod.forward)(*args, **kwargs)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 212, in _fn
    return fn(*args, **kwargs)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py", line 333, in catch_errors
    return callback(frame, cache_size, hooks)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 480, in _convert_frame
    result = inner_convert(frame, cache_size, hooks)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 103, in _fn
    return fn(*args, **kwargs)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/utils.py", line 88, in time_wrapper
    r = func(*args, **kwargs)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 339, in _convert_frame_assert
    return _compile(
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 400, in _compile
    out_code = transform_code_object(code, transform)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py", line 341, in transform_code_object
    transformations(instructions, code_options)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py", line 387, in transform
    tracer.run()
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 1684, in run
    super().run()
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 538, in run
    and self.step()
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 501, in step
    getattr(self, inst.opname)(inst)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 307, in wrapper
    return inner_fn(self, inst)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py", line 137, in impl
    self.push(fn_var.call_function(self, self.popn(nargs), {}))
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 356, in call_function
    result = handler(tx, *args, **kwargs)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/variables/builtin.py", line 608, in call_getitem
    return args[0].call_method(tx, "__getitem__", args[1:], kwargs)
  File "/extern/home/nisheethlahoti/miniconda3/envs/fitb2/lib/python3.10/site-packages/torch/_dynamo/variables/nn_module.py", line 378, in call_method
    assert type(module).__getitem__ in (
AssertionError: BufferList

from user code:
   File "/tmp/model.py", line 26, in forward
    return x + self.buflist[0]

Set torch._dynamo.config.verbose=True for more information


You can suppress this exception and fall back to eager by setting:
    torch._dynamo.config.suppress_errors = True
```

If I change the second-last line to `model = Dummy()`, I get the expected output `tensor([1., 2., 3., 4., 5.])`

### Versions

PyTorch version: 2.0.0.dev20221226
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:26:04) [GCC 10.4.0] (64-bit runtime)
Python platform: Linux-5.15.0-1029-azure-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.7.99
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 515.86.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-fid==0.2.0
[pip3] pytorch-lightning==1.8.6
[pip3] pytorch-msssim==0.2.1
[pip3] torch==2.0.0.dev20221226
[pip3] torchaudio==2.0.0.dev20221226
[pip3] torchmetrics==0.11.0
[pip3] torchvision==0.15.0.dev20221226
[conda] blas                      1.0                         mkl
[conda] libblas                   3.9.0            12_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            12_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            12_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            12_linux64_mkl    conda-forge
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0           py310h7f8727e_0
[conda] numpy                     1.23.5          py310h53a5b5f_0    conda-forge
[conda] pytorch                   2.0.0.dev20221226 py3.10_cuda11.7_cudnn8.5.0_0    pytorch-nightly
[conda] pytorch-cuda              11.7                 h67b0de4_2    pytorch-nightly
[conda] pytorch-fid               0.2.0                    pypi_0    pypi
[conda] pytorch-lightning         1.8.6                    pypi_0    pypi
[conda] pytorch-msssim            0.2.1                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch-nightly
[conda] torchaudio                2.0.0.dev20221226     py310_cu117    pytorch-nightly
[conda] torchmetrics              0.11.0             pyhd8ed1ab_0    conda-forge
[conda] torchtriton               2.0.0+0d7e753227           py310    pytorch-nightly
[conda] torchvision               0.15.0.dev20221226     py310_cu117    pytorch-nightly

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @mlazos @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

**Comment by ezyang** [comment]
For whoever picks this up, it would be nice if whatever solution we do works uniformly across ALL magic methods for all variable trackers, not just a one off fix for Module.

**Comment by yanboliang** [comment]
This has been fixed by #97932, where we inlining the custom ```__getitem__``` function. I'll close this and feel free to reopen it if not work.
